### PR TITLE
Core: Add has_list and count_list and improve (?) count_group

### PR DESCRIPTION
**Pull request by NewSoupVi**
### has_list and count_list
`has_group` and `count_group` currently exist to be able to check for a specific overall quantity of a list of items.
However, these need item groups to be defined, and also the user has no way of telling this function **which items from a group actually exist**. Those functions will just blindly try all items from the item group.

So, I decided to add two new functions called `has_list` and `count_list`, that are completely analogous to `has_group` and `count_group`, except you can specify the list of items you care about. This way, you 1. don't need an item group to be defined for the items and 2. can prepare the item list yourself, first sorting out any items that don't exist because of e.g. the options the player chose.

### count_group refactor

This function uses a for loop for a sum despite not having any sort of early exit / lazy eval (like e.g. `has_group` does). This should probably just be a `return sum(...)` expression.

## Motivation

[An new world dev](https://discord.com/channels/731205301247803413/1214608557077700720/1216166637884014662) asked for it and [me](https://discord.com/channels/731205301247803413/1214608557077700720/1216167608324198521) and [Treble](https://discord.com/channels/731205301247803413/1214608557077700720/1216167583057973399) agreed that the current implementation is not really [das Gelbe vom Ei](https://en.wiktionary.org/wiki/das_Gelbe_vom_Ei), at least if it's going to be the only option for this.

## Considerations

Some of these decisions might have been made because of how frozen builds are optimised. I'm not sure how to judge or test that. If the `count_group` refactor actually makes the frozen build slower bc of some weird optimisation thing, I am happy to revert it and take it out of the scope of this PR.

### CODE REVIEW

**Review by Ixrec — APPROVED** [review]
Since `has_list` and `has_group` appear to have identical implementations except for the `self.multiworld.worlds[player].item_name_groups[item_name_group]` expression, maybe we should deduplicate them by reimplementing `has_group` as a call to `has_list`?

Code LGTM and I agree this is useful. Methods are short enough it's fine if we leave them slightly duped.

Can't help with the concerns about frozen performance.

**Comment by NewSoupVi** [comment]
> Since `has_list` and `has_group` appear to have identical implementations except for the `self.multiworld.worlds[player].item_name_groups[item_name_group]` expression, maybe we should deduplicate them by reimplementing `has_group` as a call to `has_list`?

I try to avoid calling functions in state methods because function calls take (an admittely very tiny amount of) time, duplicating the code is faster. At least in bare python ,compiled Python might be different. But even if it does take time, in this case I think it could maybe be worth it, as this is a slower function so adding one function call would probably not be the worst thing. Would like more opinions on that if ppl have them

**Review by Exempt-Medic — APPROVED** [review]
Changes LGTM. Tested them with a few groups that they actually worked

**Comment by NewSoupVi** [comment]
Making a note here: If both this PR and https://github.com/ArchipelagoMW/Archipelago/pull/3192 are merged, that implies the necessity of two more new functions, `has_list_exclusive` and `count_list_exclusive`

File: BaseClasses.py [hunk 1]
```diff
@@ -767,6 +767,18 @@ def item_count(self, item: str, player: int) -> int:
         Utils.deprecate("Use count instead.")
         return self.count(item, player)
 
+    def has_list(self, items: Iterable[str], player: int, count: int = 1) -> bool:
```
Comment (beauxq) [inline BaseClasses.py:770]: This behavior doesn't really match what I would expect from this function name.
With this name, I would expect it to make sure I have everything in the list.

maybe `has_from_list` (and `count_from_list`)?

Comment (NewSoupVi) [reply BaseClasses.py:770]: What do you think about the existing `has_group` and `count_group` that do the same thing as these new functions?

Comment (NewSoupVi) [reply BaseClasses.py:770]: Oh ftr I agree with you, but I'm not sure if this is a can I wanna open

Comment (beauxq) [reply BaseClasses.py:770]: Yeah, I didn't think about comparing it with `has_group` and `count_group` from the same perspective.
It does seem like the same reasoning would apply to those, but it's also more troublesome to change them.

At least if they don't change, I think it would be good for them (all 4) to have doc strings that make it clear that it's not the other interpretation.

Comment (NewSoupVi) [reply BaseClasses.py:770]: I added some docstrings

File: BaseClasses.py [hunk 2]
```diff
@@ -767,8 +767,25 @@ def item_count(self, item: str, player: int) -> int:
...
+        """Returns True if the cumulative count of items from a list present in state exceeds the specified number."""
```
Comment (beauxq) [inline BaseClasses.py:771]: The way I would interpret it:
"exceeds" `>`
"meets or exceeds" `>=`

But that also makes it really wordy, and I'm trying to think of a way to make it less wordy...

File: BaseClasses.py (hunk 1 above)
Comment (black-sliver) [reply BaseClasses.py:770]: I think `has_from_list` is either easier to understand, or if not, it would at least give more reason to actually look at the doc string.

Typing-wise I'd have to read up on, but maybe Container is more accurate than Iterable, because we only need the `in` operator (`__contains__` method).

File: BaseClasses.py (hunk 2 above)
Comment (black-sliver) [reply BaseClasses.py:771]: just "meets" would be fine, i think, because "too many" is not a thing for state.has.

Comment (Exempt-Medic) [reply BaseClasses.py:771]: I think "meets" is also misleading, since people might assume it must be exactly equal to that value

File: BaseClasses.py [hunk 3]
```diff
@@ -767,8 +767,25 @@ def item_count(self, item: str, player: int) -> int:
         Utils.deprecate("Use count instead.")
         return self.count(item, player)
 
+    def has_list(self, items: Iterable[str], player: int, count: int = 1) -> bool:
```
Comment (Exempt-Medic) [inline BaseClasses.py:770]: I'm wondering if this should remove the default of `1` since, as far as I can tell, if you're actually using `1` you should just be using `has_any` instead

File: BaseClasses.py (hunk 2 above)
Comment (beauxq) [reply BaseClasses.py:771]: How about this as the doc string?
```
Has `player` collected at least `count` items matching `items`?
```

File: BaseClasses.py (hunk 1 above)
Comment (NewSoupVi) [reply BaseClasses.py]: This is `in` as in `for a in b`, not `in` as in `if a in b`. Those are different operators. The former requires `Iterable`.

Comment (NewSoupVi) [reply BaseClasses.py]: Renamed to `has_from_list` and `count_from_list`

File: BaseClasses.py (hunk 2 above)
Comment (NewSoupVi) [reply BaseClasses.py]: I like that conceptually, but I'll reword it to not be a question

**Comment by NewSoupVi** [comment]
Addressed all the feedback

Rewritten docstrings should be rereviewed

File: BaseClasses.py (hunk 1 above)
Comment (black-sliver) [reply BaseClasses.py]: > This is in as in for a in b, not in as in if a in b. Those are different operators. The former requires Iterable.

Yeah, I misread the code and realized this made no sense after turning off the PC for the night :sweat:

**Review by black-sliver — APPROVED** [review]
Code and commends look correct and same as with the other PR: moving more things to state allows us to change implementations without changing worlds.

File: BaseClasses.py [hunk 4]
```diff
@@ -767,8 +767,23 @@ def item_count(self, item: str, player: int) -> int:
         Utils.deprecate("Use count instead.")
         return self.count(item, player)
 
+    def has_from_list(self, items: Iterable[str], player: int, count: int) -> bool:
```
Comment (Jarno458) [inline BaseClasses.py:770]: to be honest, the doc string do make it clear, but when i see `state.has_from_list(["itemA", "ItemB"], 3, player)` it would be unclear to me at first glance what it does
`has_from_list` to me from its method name looks very simular to `state.has_any(["itemA", "ItemB"], player)`
so my recommendation would be to name it `has_atleast()`

File: BaseClasses.py [hunk 5]
```diff
@@ -767,8 +767,23 @@ def item_count(self, item: str, player: int) -> int:
...
+        """Returns True if the state contains at least `count` items matching any of the item names from a list."""
+        found: int = 0
+        player_prog_items = self.prog_items[player]
+        for item_name in items:
+            found += player_prog_items[item_name]
+            if found >= count:
+                return True
+        return False
+
+    def count_from_list(self, items: Iterable[str], player: int) -> int:
```
Comment (Jarno458) [inline BaseClasses.py:780]: why not just name this `count` and make it an overload?

**Review by Jarno458 — CHANGES_REQUESTED** [review]
I dont see any tests added for these new state checks
